src/interface/swarm.cpp

```diff
@@ -317,28 +354,469 @@ void Swarm::Defrag() {
+  PackIndexMap rmap;
+  PackIndexMap imap;
+  PackAllVariables(vreal, vint, rmap, imap);
+  int real_vars_size = realVector_.size();
+  int int_vars_size = intVector_.size();
+
+  pmb->par_for(
+      "Swarm::DefragVariables", 0, max_active_index_, KOKKOS_LAMBDA(const int n) {
+        if (from_to_indices(n) >= 0) {
+          for (int i = 0; i < real_vars_size; i++) {
+            vreal(i, from_to_indices(n)) = vreal(i, n);
           }
-        });
+          for (int i = 0; i < int_vars_size; i++) {
+            vint(i, from_to_indices(n)) = vint(i, n);
+          }
+        }
+      });
+
+  // Update max_active_index_
+  max_active_index_ = num_active_ - 1;
+}
+
+void Swarm::SetupPersistentMPI() {
+  vbswarm->SetupPersistentMPI();
+
+  allreduce_request_ = MPI_REQUEST_NULL;
+
+  // Index into neighbor blocks
+  auto pmb = GetBlockPointer();
+  auto neighborIndices_h = neighborIndices_.GetHostMirror();
+
+  // Region belonging to this meshblock
+  // for (int k = 1; k < 3; k++) {
+  for (int k = 0; k < 4; k++) { // 2D ONLY!
+    for (int j = 1; j < 3; j++) {
+      for (int i = 1; i < 3; i++) {
+        neighborIndices_h(k, j, i) = this_block_;
+      }
+    }
+  }
+
+  // TODO(BRR) Checks against some current limitations
+  int ndim = pmb->pmy_mesh->ndim;
+  PARTHENON_REQUIRE(ndim == 2, "Only 2D tested right now!");
+  auto mesh_bcs = pmb->pmy_mesh->mesh_bcs;
+  for (int n = 0; n < 2 * ndim; n++) {
+    PARTHENON_REQUIRE(mesh_bcs[n] == BoundaryFlag::periodic,
+                      "Only periodic boundaries supported right now!");
+  }
+
+  for (int n = 0; n < pmb->pbval->nneighbor; n++) {
+    NeighborBlock &nb = pmb->pbval->neighbor[n];
+
+    int i = nb.ni.ox1;
+    int j = nb.ni.ox2;
+    // int k = nb.ni.ox3;
+
+    if (ndim == 1) {
+      if (i == -1) {
+        neighborIndices_h(0, 0, 0) = n;
+      } else if (i == 0) {
+        neighborIndices_h(0, 0, 1) = n;
+        neighborIndices_h(0, 0, 2) = n;
+      } else {
+        neighborIndices_h(0, 0, 3) = n;
+      }
+    } else if (ndim == 2) {
+      if (i == -1) {
+        if (j == -1) {
+          neighborIndices_h(0, 0, 0) = n;
+        } else if (j == 0) {
+          neighborIndices_h(0, 1, 0) = n;
+          neighborIndices_h(0, 2, 0) = n;
+        } else if (j == 1) {
+          neighborIndices_h(0, 3, 0) = n;
+        }
+      } else if (i == 0) {
+        if (j == -1) {
+          neighborIndices_h(0, 0, 1) = n;
+          neighborIndices_h(0, 0, 2) = n;
+        } else if (j == 1) {
+          neighborIndices_h(0, 3, 1) = n;
+          neighborIndices_h(0, 3, 2) = n;
+        }
+      } else if (i == 1) {
+        if (j == -1) {
+          neighborIndices_h(0, 0, 3) = n;
+        } else if (j == 0) {
+          neighborIndices_h(0, 1, 3) = n;
+          neighborIndices_h(0, 2, 3) = n;
+        } else if (j == 1) {
+          neighborIndices_h(0, 3, 3) = n;
+        }
+      }
+    } else {
+      PARTHENON_FAIL("3D particles not currently supported!");
```

Comment:
What are the blockers for 3D particles?

Reply:
I *think* just establishing neighbor indices in 3D (and 1D). And then testing that it works. Maybe I'll just go ahead and do that now, I might run into a problem though.